src/components/implementation/no_interface/cFE_booter/osfilesys.c

```diff
@@ -1,280 +1,886 @@
+
+            // bytes_to_read < the continuous space left on f_part
+            } else {
+                memcpy(buffer, position->open_part->data + position->part_offset, bytes_to_read);
+                position->part_offset     += bytes_to_read;
+                position->file_offset     += bytes_to_read;
+                return nbytes;
+            }
+        }
+    } else if (o->memtype == STATIC) {
+        memcpy(buffer, &part->data[position->file_offset], bytes_to_read);
+        position->part_offset             += bytes_to_read;
+        position->file_offset             += bytes_to_read;
+        return nbytes;
+    } else {
+        PANIC("Memtype is neither static nor dynamic");
+    }
+
+    PANIC("Unreachable Statement");
+    return 0;
+}
+
+int32 file_write(int32 FD, void *buffer, uint32 nbytes)
+{
+    if (!buffer) return OS_FS_ERR_INVALID_POINTER;
+    int32 ret = chk_fd(FD);
+    if (ret != OS_FS_SUCCESS) return ret;
+
+    struct fd *filedes = &fd_tbl[FD];
+    struct fsobj *o = filedes->file;
+    struct file_position *position = &filedes->position.file_pos;
+
+    if (o->refcnt < 1)                  return OS_FS_ERROR;
+    if (o->memtype == STATIC)           return OS_FS_ERROR;
+    if (o->type == FSOBJ_DIR)           return OS_FS_ERROR;
+
+    if (nbytes == 0)                    return 0;
+    uint32 bytes_to_write = nbytes;
+    uint32 bytes_remaining = F_PART_DATA_SIZE - position->part_offset;
+
+    //while there are enough bytes to be written to fill a f_part
+    while (bytes_to_write > bytes_remaining) {
+        memcpy(position->open_part->data + position->part_offset, buffer, bytes_remaining);
+        position->file_offset           += bytes_remaining;
+        buffer                          += bytes_remaining;
+        bytes_to_write                  -= bytes_remaining;
+        position->part_offset           = 0;
+        if (position->open_part->next == NULL) {
+            struct f_part *part;
+            newpart_get(&part);
+            part->file                  = o;
+            part->next                  = NULL;
+            part->prev                  = position->open_part;
+            position->open_part->next   = part;
+        }
+
+        position->open_part             = position->open_part->next;
+        bytes_remaining                 = F_PART_DATA_SIZE - position->part_offset;
+    }
+    //bytes_to_write < bytes_remaining
+    memcpy(position->open_part->data, buffer, bytes_to_write);
+    position->part_offset               += bytes_to_write;
+    position->file_offset               += bytes_to_write;
+    if (o->size < position->file_offset) {
+        o->size                         = position->file_offset;
+    }
+    return nbytes;
 }
 
-uint32 file_insert(struct fsobj *o, struct FS *filesys)
+struct fsobj *file_find(char *path)
 {
-    assert(filesys && filesys->root && filesys->root->filedes);
-    if (filesys->root == o) return OS_FS_SUCCESS;
+    assert(path);
```

Comment:
It seems logical to return a null to a function that returns a file if that file cannot be found.  Idk, does that make sense, or is there something i'm missing